debug: openocd: Fix building for SMP platforms

abrodkin opened the pull request:
struct k_thread *current is only defined for !SMP platforms
leading to the following build-time failure:

| openocd.c:39:35: error: 'struct z_kernel' has no member named 'current'
|  [OPENOCD_OFFSET_K_CURR_THREAD] = offsetof(struct z_kernel, current),
                                   ^~~~~~~~

Disable calculation of meaningless offset then.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/18124.

galak commented on subsys/debug/openocd.c:
```diff
@@ -36,7 +36,9 @@ __attribute__((used, section(".openocd_dbg")))
 size_t _kernel_openocd_offsets[] = {
 	/* Version 0 starts */
 	[OPENOCD_OFFSET_VERSION] = 1,
+#ifndef CONFIG_SMP
```
is this going to cause issues w/openocd and the assumed layout of this data?  Wondering if setting the value to 0 is better?

ruuddw replied:
0 is probably better, though might break things as well (don't know if openocd/gdb zephyr awareness relies on this).
Openocd Zephyr-SMP support might require more. Maybe instead of working around this, do not allow/use combination of SMP and openocd zephyr awareness for now?

MaureenHelm replied:
`OPENOCD_UNIMPLEMENTED` would be better than zero, but I also like @abrodkin's idea to handle this in KConfig.

abrodkin commented:
@galak @ruuddw so maybe something like this until we really make sure OpenOCD works with Zephyr-aware on SMP systems?
```
diff --git a/subsys/debug/Kconfig b/subsys/debug/Kconfig
index a55d0f3350..a5dff0ad4f 100644
--- a/subsys/debug/Kconfig
+++ b/subsys/debug/Kconfig
@@ -223,6 +223,7 @@ config EXCEPTION_STACK_TRACE

 config OPENOCD_SUPPORT
        bool "OpenOCD support [EXPERIMENTAL]"
+       depends on !SMP
        select THREAD_MONITOR
        select THREAD_NAME
        help
```

ioannisg commented:
@nashif @galak @ruuddw could you finalize reviewing this bug-fix of a a high priority bug?

ioannisg commented:
Moving to next release since the review was not finalized for v.2.0.0

galak commented:
Any updates on this?

abrodkin commented:
@EvgeniiDidin could you please take a look at this one since you anyways deal with OpenOCD/GDB thingy these days?

galak commented:
@abrodkin, @EvgeniiDidin  any update on this?

EvgeniiDidin commented:
Hi, 

I tested SMP Zephyr smp/pi example with these changes on HSDK(4 core) board using OpenOCD/ gdb. It is possible to load and start the application on all cores, also I was able to get callee registers from stack. But after stopping execution "continue" command failed with next message:
```
(gdb) c
Cannot execute this command without a live selected thread.
```
I guess it is related to the lack of  current_thread{id} values in OpenOCD, which come from "current" struct. 

If  incomplete smp aware support is permissible for Zephyr/OpenOCD,  this commit is fine. Otherwise we should disable  simultaneous usage of CONFIG_SMP and CONFIG_OPENOCD options.

galak commented:
@abrodkin, @EvgeniiDidin any update on this?

abrodkin commented:
@galak I think the most recent comment of Eugeniy (https://github.com/zephyrproject-rtos/zephyr/pull/18125#issuecomment-559171331) still valid. We either declare SMP mode doesn't support additional OpenOCD-related visibility options or ignore pointer to the "current thread" implemented in this commit. I now lean towards making `CONFIG_OPENOCD` incompatible with `CONFIG_SMP` as at least it won't introduce any funky behavior.

ruuddw commented:
Let's prevent openocd and SMP / multiple CPUs in kconfig.
SMP/multi-core is simply not yet supported. See also PR #23503 : adds a #error in openocd.c.

abrodkin committed:
OpenOCD: Make incompatible with SMP

In case of SMP (i.e. multiple execution units processing
the same list of tasks) we cannot use the same data structures for
getting data about active tasks as with just one processor (UP).

So until explicit support of SMP is added make sure we don't allow
to select both OPENOCD & SMP simultaneously.

Moreover starting from
commit a203d21962e7 ("kernel: remove legacy fields in _kernel")
this will lead to build-time error if MP_NUM_CPUS > 1.

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

carlescufi reviewed (approved):
LGTM but I need @andrewboie, @MaureenHelm, @galak or @nashif  to approve
